### PR TITLE
Do not return on hold judge tasks to the front end

### PR DESCRIPTION
**Pull request by hschallhorn**
Hopefully resolves inability to load a judge queue.

### Description
Judge queues do not have an on hold or completed tab. Do not send these tasks to the front end. Also, we individually load attorney queues and tasks when the user clicks into their queue, so do not include relevant attorney tasks.

Proof is in the ~pudding~ `Benchmark.measure`
```ruby
user = User.third
tasks = QueueForRole.new("judge").create(user: user).tasks; nil
Benchmark.measure do
  AmaAndLegacyTaskSerializer.create_and_preload_legacy_appeals(
      params: { user: user, role: "judge" },
      tasks: tasks,
      ama_serializer: WorkQueue::TaskSerializer
    ).call
end
# master
=> #<Benchmark::Tms:0x00007fa536f01760 @cstime=0.0, @cutime=0.0, @label="", @real=10.44135700003244, @stime=0.3461829999999999, @total=6.489523999999999, @utime=6.1433409999999995>
# hschallhorn/judge-queue-fix
=> #<Benchmark::Tms:0x00007fa5346d6ba0 @cstime=0.0, @cutime=0.0, @label="", @real=3.1014509999658912, @stime=0.14628300000000038, @total=2.2935359999999996, @utime=2.147252999999999>
```

Benchmarked for user in prod
```ruby
user = User.find_by_css_id("bvampeters".upcase)
tasks = QueueForRole.new("judge").create(user: user).tasks; nil
Benchmark.measure do
  AmaAndLegacyTaskSerializer.create_and_preload_legacy_appeals(
      params: { user: user, role: "judge" },
      tasks: tasks,
      ama_serializer: WorkQueue::TaskSerializer
    ).call
end
=> #<Benchmark::Tms:0x0000000012015e48 @label="", @real=65.12194121100038, @cstime=0.0, @cutime=0.0, @stime=1.081439, @utime=33.623063, @total=34.704502000000005>
tasks = Task.incomplete_or_recently_closed.active.where(assigned_to: user).includes(*[
        { appeal: [:available_hearing_locations, :claimants] },
        :assigned_by,
        :assigned_to,
        :children,
        :parent
      ]).order(created_at: :asc).limit(10_000)
=> #<Benchmark::Tms:0x0000000011c37268 @label="", @real=8.98823658100082, @cstime=0.0, @cutime=0.0, @stime=0.23679000000000006, @utime=6.75309, @total=6.98988>
```
#### Loading a judge review page
1. Log in as bvaaabshire
1. On master load bvaaabshire's queue and note the get tasks request returns 113ish tasks even though the queue only displays 28 tasks
1. Check out this branch and try again
1. Note that while the number of tasks in the user's queue is unchanged only 11ish tasks were sent to the front end

BEFORE|AFTER
---|---
Screen Shot 2020-04-27 at 7 06 26 PM|Screen Shot 2020-04-28 at 3 10 41 PM
#### Loading an attorney's queue
1. Log in as bvaaabshire
1. On master load bvaaabshire's assign queue and note the get tasks request returns 113ish tasks even though the queue only displays 28 tasks
1. Click into an attorney's queue and ensure the same number of tasks load
1. Check out this branch and try again
1. Note that while the number of tasks in the user's queue is unchanged only 11ish tasks were sent to the front end
1. Click into an attorney's queue and ensure the same number of tasks load

### User Facing Changes
 - [ ] Load times 🤞

### CODE REVIEW

File: app/models/queues/generic_queue.rb [hunk 1]
```diff
@@ -7,33 +7,21 @@ def initialize(limit: 10_000, user:)
...
+    relevant_tasks
   end
 
   private
 
   attr_reader :limit, :user
 
   def relevant_tasks
-    Task.incomplete_or_recently_closed
+    Task.incomplete_or_recently_closed.visible_in_queue_table_view
```
Comment (hschallhorn) [inline app/models/queues/generic_queue.rb:18]: Don't send timed hold tasks to the front end. They will not be shown in user's queues.

File: app/models/queues/generic_queue.rb [hunk 2]
```diff
@@ -7,33 +7,21 @@ def initialize(limit: 10_000, user:)
...
-    return [] unless user.is_a?(User)
-
-    # If the user is a judge there will be attorneys in the list, if the user is not a judge the list of attorneys will
-    # be an empty set and this function will also then return an empty set.
-    AttorneyTask.incomplete_or_recently_closed
-      .where(assigned_to: Judge.new(user).attorneys)
-      .includes(*task_includes)
-      .order(created_at: :asc)
-      .limit(limit)
-  end
```
Comment (hschallhorn) [inline app/models/queues/generic_queue.rb:35]: Moved to judge queue

File: app/models/queues/judge_queue.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+class JudgeQueue < GenericQueue
+  def tasks
+    (relevant_tasks.active + relevant_attorney_tasks)
```
Comment (hschallhorn) [inline app/models/queues/judge_queue.rb:5]: Only show assigned or in-progress tasks assigned to the user, in addition to their attorneys' tasks

File: spec/controllers/tasks_controller_spec.rb [hunk 4]
```diff
@@ -139,7 +140,8 @@
         expect(in_progress["attributes"]["placed_on_hold_at"]).to be nil
 
         # Ensure we include recently completed tasks
-        expect(response_body.count { |task| task["id"] == task10.id.to_s }).to eq 1
+        expect(response_body.count { |task| task["id"] == task10.id.to_s }).to eq 0
+        expect(response_body.count { |task| task["id"] == task16.id.to_s }).to eq 0
```
Comment (hschallhorn) [inline spec/controllers/tasks_controller_spec.rb:144]: no on hold or completed tasks are returned.

File: app/models/task.rb [hunk 5]
```diff
@@ -60,6 +60,10 @@ class << self; undef_method :open; end
 
   scope :not_cancelled, -> { where.not(status: Constants.TASK_STATUSES.cancelled) }
 
+  scope :recently_closed, -> { closed.where(closed_at: (Time.zone.now - 1.week)..Time.zone.now) }
+
+  scope :incomplete_or_recently_closed, -> { open.or(recently_closed) }
+
```
Comment (lomky) [inline app/models/task.rb:66]: Great change!

File: app/workflows/queue_for_role.rb [hunk 6]
```diff
@@ -1,6 +1,11 @@
 # frozen_string_literal: true
 
 class QueueForRole
+  QUEUE_ROLE_MAPPER = {
+    attorney: AttorneyQueue,
+    judge: JudgeQueue
+  }
```
Comment (lomky) [inline app/workflows/queue_for_role.rb:7]: :echo-awe:

File: app/controllers/tasks_controller.rb [hunk 7]
```diff
@@ -74,9 +74,9 @@ def create
     param_groups.each do |task_type, param_group|
       tasks << valid_task_classes[task_type.to_sym].create_many_from_params(param_group, current_user)
     end
-    tasks.flatten!
+    modified_tasks = [parent_tasks_from_params, tasks].flatten!
```
Comment (hschallhorn) [inline app/controllers/tasks_controller.rb:77]: Ensures the updated on hold parent is returned as queue for role will no longer return on hold tasks for judges. Ensures that when a judge adds an admin action to a task, the parent that has been updated to on hold is returned to the front end so the task can be removed from the users queue.

File: spec/controllers/tasks_controller_spec.rb [hunk 8]
```diff
@@ -370,6 +370,60 @@
...
+          it "should be successful" do
+            expect(AppealRepository).to receive(:update_location!).exactly(2).times
+
+            subject
+
+            expect(response.status).to eq 200
+            response_body = JSON.parse(response.body)["tasks"]["data"]
+            expect(response_body.size).to eq(5)
+
+            # Ensure the parent task is also returned
```
Comment (lomky) [inline spec/controllers/tasks_controller_spec.rb:401]: nit: I'd love a comment per section saying what we're expecting to be interrogating in the first, second, etc

File: app/models/queues/generic_queue.rb (hunk 2 above)
Comment (hschallhorn) [reply app/models/queues/generic_queue.rb:35]: Actually removed all together. We load individual queues when the user actually [clicks into the attorney queue](https://github.com/department-of-veterans-affairs/caseflow/pull/12206/files#r333239891)

File: app/controllers/tasks_controller.rb [hunk 9]
```diff
@@ -149,7 +149,7 @@ def queue_config
   def verify_view_access
     return true unless FeatureToggle.enabled?(:scm_view_judge_assign_queue)
 
-    return true if user == current_user
+    return true if user == current_user || Judge.new(current_user).attorneys.include?(user)
```
Comment (hschallhorn) [inline app/controllers/tasks_controller.rb:152]: Allows the user to request the tasks of their attorneys when SCM is enabled. Before this change, we were getting and populating attorney tasks from the original request for the judge's tasks. This request was failing, but did not matter as we already had the tasks on the front end. We should continue to do this request and not initially load attorney tasks to further improve the performance of judge queues.

File: app/models/queues/judge_queue.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class JudgeQueue < GenericQueue
+  def tasks
+    super.active
```
Comment (hschallhorn) [inline app/models/queues/judge_queue.rb:5]: Do not send on hold or completed tasks to the front end

File: spec/controllers/tasks_controller_spec.rb [hunk 11]
```diff
@@ -138,8 +142,10 @@
         expect(in_progress["attributes"]["assigned_to"]["id"]).to eq user.id
         expect(in_progress["attributes"]["placed_on_hold_at"]).to be nil
 
-        # Ensure we include recently completed tasks
-        expect(response_body.count { |task| task["id"] == task10.id.to_s }).to eq 1
+        # Ensure we don't include recently completed tasks, on hold tasks, or attorney tasks
+        expect(response_body.count { |task| task["id"] == task10.id.to_s }).to eq 0
+        expect(response_body.count { |task| task["id"] == task16.id.to_s }).to eq 0
+        expect(response_body.count { |task| task["id"] == task17.id.to_s }).to eq 0
```
Comment (hschallhorn) [inline spec/controllers/tasks_controller_spec.rb:148]: No attorney tasks are returned

**Pull request merged**